studio/src/hooks/useKeySequence.ts

```diff
@@ -0,0 +1,89 @@
+import { useInputFocusDetection } from "@/hooks";
+import { useCallback, useEffect, useRef, useState } from "react";
+
+type KeySequenceOptions = {
+  enabled?: boolean;
+  description?: string;
+  timeout?: number;
+};
+
+export function useKeySequence(
+  sequence: string[],
+  callback: () => void,
+  options?: KeySequenceOptions,
+) {
+  const { enabled = true, timeout = 2000 } = options ?? {};
+
+  const { isInputFocused } = useInputFocusDetection();
+
+  const [ref, setRef] = useState<HTMLElement | null>(null);
+  const [keySequence, setKeySequence] = useState<string[]>([]);
+
+  const timeoutRef = useRef<number | NodeJS.Timeout | undefined>(undefined);
+  const callbackRef = useRef(callback);
+  const sequenceRef = useRef(sequence);
+
+  // ensure that the callback and sequence are updated when they change
+  useEffect(() => {
+    sequenceRef.current = sequence;
+  }, [sequence]);
+
+  useEffect(() => {
+    callbackRef.current = callback;
+  }, [callback]);
+
+  // create and persist the callback across re-renders
+  const handleKeyPress = useCallback((event: Event) => {
+    if (!(event instanceof KeyboardEvent)) {
+      return;
+    }
+
+    setKeySequence((prevKeySequence) => {
+      const updatedSequence = [...prevKeySequence, event.key].slice(
+        -sequenceRef.current.length,
+      );
+
+      return updatedSequence;
+    });
+  }, []);
+
+  useEffect(() => {
+    if (keySequence.join("") === sequenceRef.current.join("")) {
```

Comment:
Can you add some comments here. Why are we using a reference for sequence? That should not matter, right? I'm guessing its done because the hook gets called with an array of keys (sequence) and that array may be recreated each render.